Explore primary constructors

BillWagner opened the pull request:
Add a tutorial for the C# 12 feature primary constructors.

Fixes #34529

I went through all the comments on the [Preview feedback discussion](https://github.com/dotnet/csharplang/discussions/7109) for primary constructors. This tutorial provides examples that reinforce the LDM decisions on primary constructors, their parameters and how they are modeled.


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/whats-new/csharp-12.md](https://github.com/dotnet/docs/blob/018884baa5884486fe4480fbfcc949ca8716db85/docs/csharp/whats-new/csharp-12.md) | [What's new in C# 12](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12?branch=pr-en-us-35742) |
| [docs/csharp/whats-new/tutorials/primary-constructors.md](https://github.com/dotnet/docs/blob/018884baa5884486fe4480fbfcc949ca8716db85/docs/csharp/whats-new/tutorials/primary-constructors.md) | [docs/csharp/whats-new/tutorials/primary-constructors](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors?branch=pr-en-us-35742) |
| [docs/csharp/whats-new/tutorials/records.md](https://github.com/dotnet/docs/blob/018884baa5884486fe4480fbfcc949ca8716db85/docs/csharp/whats-new/tutorials/records.md) | [Create record types](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/records?branch=pr-en-us-35742) |


<!-- PREVIEW-TABLE-END -->

BillWagner committed:
create outline

Capture notes from the [primary constructors preview discussion](https://github.com/dotnet/csharplang/discussions/7109)

BillWagner committed:
interim checkin

BillWagner committed:
add primary constructor tutorial

Explore primary constructors.

BillWagner committed:
one small edit

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+Key learning experiences for this tutorial:
+
+1. declare a primary constructor when parameters are needed in one or more members.
+    1. initialize properties (same or different type.)
```
also fields

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+    1. initialize properties (same or different type.)
+    1. as a service or dependencies.
+    1. pass to base class with primary constructor
+1. consider when a reasonable can be used in a default ctor.
```
i'd recommend "constructor" not "ctor"

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+
+## Primary constructors
+
+You can add parameters to a `struct` or `class` declaration to create a *primary constructor*. You declare a primary constructor when a struct or class **requires** parameters to initialize an instance. Primary constructor parameters are in scope throughout the class definition. It's important to view primary constructor parameters as *parameters* even though they are in scope throughout the clas ...
```
not sure i understand second sentence.  feels like it could removed.

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+
+You can add parameters to a `struct` or `class` declaration to create a *primary constructor*. You declare a primary constructor when a struct or class **requires** parameters to initialize an instance. Primary constructor parameters are in scope throughout the class definition. It's important to view primary constructor parameters as *parameters* even though they are in scope throughout the clas ...
+
+1. Primary constructor parameters aren't stored directly as fields.
```
Instead of saying that they're not stored directly as fields (which they are in the roslyn impl), i'd say something more like "how parameters are stored isn't specified and compilers are free to store or not store them however they want".

I'd also add "primary constructor parameters may not be stored at all if not needed".

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+
+1. Primary constructor parameters aren't stored directly as fields.
+1. Primary constructors aren't members of the class. For example, a primary constructor parameter named `param` can't be accessed as `this.param`.
+1. Primary constructor parameters are mutable.
```
I'd say "can be assigned to"

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+The two most common uses for a primary constructor parameter are:
+
+1. As an argument to a `base()` constructor invocation.
+1. To initialize a member field or property.
```
i'd say there are three common usages.  the two above, and also just referencing the parameter directly from within members of hte type.

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,110 @@
+1. As an argument to a `base()` constructor invocation.
+1. To initialize a member field or property.
+
+Every constructor for a class **must** call the primary constructor, directly or indirectly, through a `this()` constructor invocation. That rule ensures that primary constructor parameters are assigned anywhere in the body of the class.
```
"Every other constructor" perhaps?

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/snippets/primary-constructors/BankAccount.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+        CurrentBalance -= amount;
+    }
+    public override string ToString() => $"{base.ToString()}, Balance: {CurrentBalance}";
```
```suggestion

    public override string ToString() => $"{base.ToString()}, Balance: {CurrentBalance}";
```

CyrusNajmabadi commented on docs/csharp/whats-new/tutorials/snippets/primary-constructors/primary-constructors.csproj:
```diff
@@ -0,0 +1,16 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <RootNamespace>primary_constructors</RootNamespace>
```
generally, our namespaces shoudl be PascalCased.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Cyrus Najmabadi <cyrus.najmabadi@gmail.com>

BillWagner committed:
respond to feedback

BillWagner committed:
add link from whats new page.

BillWagner committed:
typo

CamSoper reviewed (approved):
I have nothing to add, LGTM!

pkulikov commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,102 @@
+You can add parameters to a `struct` or `class` declaration to create a *primary constructor*. Primary constructor parameters are in scope throughout the class definition. It's important to view primary constructor parameters as *parameters* even though they are in scope throughout the class definition. Several rules clarify that they're parameters:
+
+1. Primary constructor parameters may not be stored if they aren't needed.
+1. Primary constructors aren't members of the class. For example, a primary constructor parameter named `param` can't be accessed as `this.param`.
```
```suggestion
1. Primary constructor parameters aren't members of the class. For example, a primary constructor parameter named `param` can't be accessed as `this.param`.
```

pkulikov commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,102 @@
+1. To initialize a member field or property.
+1. Referencing the constructor parameter in an instance member.
+
+Every other constructor for a class **must** call the primary constructor, directly or indirectly, through a `this()` constructor invocation. That rule ensures that primary constructor parameters are assigned anywhere in the body of the class.
```
```suggestion
Every other constructor for a class **must** call the primary constructor, directly or indirectly, through a `this()` constructor invocation. That rule ensures that primary constructor parameters are assigned anywhere in the body of the type.
```

pkulikov commented on docs/csharp/whats-new/tutorials/snippets/primary-constructors/BankAccount.cs:
```diff
@@ -0,0 +1,115 @@
+﻿namespace primary_constructors;
+
+// <BaseClass>
+public class BankAccount(string accountID, string owner)
```
Should tutorial say anything about validation of constructor arguments, which is common practice, I assume. Do primary constructors and argument validation play together somehow?

pkulikov replied:
@BillWagner I like your additions in that last commit, thank you!

BillWagner committed:
Update docs/csharp/whats-new/tutorials/primary-constructors.md

Co-authored-by: Petr Kulikov <petr.kulikov@gmail.com>

pkulikov commented on docs/csharp/whats-new/tutorials/primary-constructors.md:
```diff
@@ -0,0 +1,102 @@
+
+It's important to understand that the first example didn't require the compiler to create a field to store the value of the primary constructor parameters. The second example used the primary constructor parameter inside a method, and therefore required the compiler create storage for them. The compiler creates storage for any primary constructors only when that parameter is accessed in the body  ...
+
+## Dependency injection
```
Maybe, that should be the first in the list of examples? The preceding two may require argument validation.

BillWagner replied:
I like the idea of adding argument validation. I've left the order as is. I think it flows better to have a simpler example first. (And, any double *should* be valid for a vector).

BillWagner committed:
Update docs/csharp/whats-new/tutorials/primary-constructors.md

Co-authored-by: Petr Kulikov <petr.kulikov@gmail.com>

BillWagner committed:
respond to feedback

Great suggestion to add examples that validate the primary constructor parameters.

BillWagner commented:
closing and reopening for a fresh build
